### ISSUE TITLE
[feat] Call javascript function from Rust (expose wry's eval_with_callback in tauri)

### ISSUE THREAD

**Issue by capvor** [tauri-apps/tauri#5441]
### Describe the problem

I recently have a project that needs to call a certain calculation function of a javascript library and get its result, which needs to be called in the rust code.

I have found the discussion https://github.com/tauri-apps/tauri/discussions/3018.
But the `window.eval("window['function_name']()")` solution can not be called with arguments and can not return data.

### Describe the solution you'd like

I would like to call a javascript function with arguments in rust code and can get its return data.

**Comment by amrbashir** [comment]
you should use the event system for this use case. I think we could implement a similar system to commands for javascript functions but it may introduce some overhead because we need to evaluate the system script so your best bet now is events.

**Comment by lucasfernog** [comment]
AFAIK there's ways to retrieve the return value from the eval call, but wry does not expose them.

**Comment by capvor** [comment]
> you should use the event system for this use case. I think we could implement a similar system to commands for javascript functions but it may introduce some overhead because we need to evaluate the system script so your best bet now is events.

Yes, currently my solution is to use the event system. It would be better if there is such a method in tauri.

**Comment by capvor** [comment]
> AFAIK there's ways to retrieve the return value from the eval call, but wry does not expose them.

Is it possible to expose them?

**Comment by MagicMajky** [comment]
This feature would greatly enhance Tauri's functionality.

I'm working on splitting and encrypting xlsx files. While I began implemnting this in Rust for performance, I discovered the Rust ecosystem lacks support for xlsx encryption. Thankfully, I found a TypeScript library that I can use for encrypting the xlsx files.

An equivalent to the command function for Rust-to-TS communication would be invaluable for my use case:

1. A user uploads a large xlsx and initiates a Rust function.
2. To manage memory efficiently, I want to process this large file in chunks, lets say 10 smaller files at a time.
3. I'd send these to TypeScript to encrypt them (and possibly saving them).
4. Once encrypted, Rust processes another batch until the large xlsx is complete.
5. Implementing this with the current event system is cumbersome:

I'd need a tracking structure, like a hashset, for pending encryptions.
Rust would emit an 'encrypt' event, handled by TS.
Post-encryption, TS would trigger a 'done' event, signaling Rust to remove the processed file from the tracking structure.
Rust would then wait for all files to be encrypted before processing the next batch.
A direct request/response mechanism between Rust and JS/TS would streamline this process immensely.

For now, I may resort to saving the small files temporarily, then encrypting them with the ts library after the rust function finishes. 
But a direct method would be ideal. Avoiding IO overhead and the complexities of the event-driven approach would be a significant win.

**Comment by liesauer** [comment]
any update?

**Comment by liesauer** [comment]
i want to execute a third-party js and get the return result, i highly prefer `let result = webviewwindow.eval("xxxxxxx")` rather than writing an event and manually pass back to rust side.

**Comment by liesauer** [comment]
another reason that i don't want to use event system is because i want all happens at the rust side since event system is base on ipc so it means that everything can be seem on the devtool panel or any network sniffer.

**Comment by FabianLars** [comment]
network sniffers shouldn't see those requests since ipc requests don't actually hit the (local) network.

since the last comments here wry gained support for return values so it "just" needs to be exposed in tauri (idk how hard that is)

**Comment by lanyeeee** [comment]
I noticed that `wry` has supported `evaluate_script_with_callback` for quite a long time https://github.com/tauri-apps/wry/pull/778 , but Tauri itself has not integrated this feature yet.

I am curious - was this feature request completely forgotten, or are there specific technical challenges preventing its implementation in Tauri? It seems like the underlying capability already exists in wry.

This would be a really valuable feature for many use cases where we need to call JavaScript functions from Rust with proper return value handling.

**Comment by FabianLars** [comment]
it was not forgotten. i'd say none of the feature requests are. we just have nobody to work on them 🤷

**Comment by liesauer** [comment]
> network sniffers shouldn't see those requests since ipc requests don't actually hit the (local) network.
> 
> since the last comments here wry gained support for return values so it "just" needs to be exposed in tauri (idk how hard that is)

those requests could be seem by enable remote debugging and forward to the network sniffer. this basically is "remote devtool" without `devtools` feature.

**Comment by lanyeeee** [comment]
I just opened a [PR](https://github.com/tauri-apps/tauri/pull/14925) that resolves this issue.
The implementation is straightforward and closely follows the existing pattern for eval_script, so it should be a quick and low-effort review.
Thanks for taking a look 🙏
